gridfs.go

```diff
@@ -0,0 +1,601 @@
+	var re *regexp.Regexp
+	var err error
+	if opts.Exclude != "" {
+		re, err = regexp.Compile(opts.Exclude)
+		if err != nil {
+			return errors.Wrap(err, "compiling exclude regex")
+		}
+	}
+
+	iter, err := b.List(ctx, opts.Remote)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	keys := []string{}
+	for iter.Next(ctx) {
+		item := iter.Item()
+		if re != nil && re.MatchString(item.Name()) {
+			continue
+		}
+
+		localName, err := filepath.Rel(opts.Remote, item.Name())
+		if err != nil {
+			return errors.Wrap(err, "getting relative filepath")
+		}
+		keys = append(keys, localName)
+
+		if err = b.Download(ctx, item.Name(), filepath.Join(opts.Local, localName)); err != nil {
+			return errors.WithStack(err)
+		}
+	}
+
+	if err = iter.Err(); err != nil {
+		return errors.WithStack(err)
+	}
+
+	if (b.opts.DeleteOnPull || b.opts.DeleteOnSync) && !b.opts.DryRun {
+		return errors.Wrap(deleteOnPull(ctx, keys, opts.Local), "deleting on sync after pull")
+	}
+
+	return nil
+}
+
+func (b *gridfsBucket) Copy(ctx context.Context, opts CopyOptions) error {
+	grip.DebugWhen(b.opts.Verbose, message.Fields{
+		"type":          "gridfs",
+		"operation":     "copy",
+		"bucket":        b.opts.Name,
+		"bucket_prefix": b.opts.Prefix,
+		"source_key":    opts.SourceKey,
+		"dest_key":      opts.DestinationKey,
+	})
+
+	from, err := b.Reader(ctx, opts.SourceKey)
+	if err != nil {
+		return errors.Wrap(err, "getting reader for source")
+	}
+
+	to, err := opts.DestinationBucket.Writer(ctx, opts.DestinationKey)
+	if err != nil {
+		return errors.Wrap(err, "getting writer for destination")
+	}
+
+	if _, err = io.Copy(to, from); err != nil {
+		return errors.Wrap(err, "copying data")
+	}
+
+	return errors.WithStack(to.Close())
+}
+
+func (b *gridfsBucket) Remove(ctx context.Context, key string) error {
+	grip.DebugWhen(b.opts.Verbose, message.Fields{
+		"type":          "gridfs",
+		"dry_run":       b.opts.DryRun,
+		"operation":     "remove",
+		"bucket":        b.opts.Name,
+		"bucket_prefix": b.opts.Prefix,
+		"key":           key,
+	})
+
+	grid, err := b.bucket(ctx)
+	if err != nil {
+		return errors.Wrap(err, "resolving bucket")
+	}
+
+	cursor, err := grid.Find(bson.M{"filename": b.normalizeKey(key)})
+	if err == mongo.ErrNoDocuments {
+		return nil
+	} else if err != nil {
+		return errors.Wrap(err, "finding file")
+	}
+
+	document := struct {
+		ID interface{} `bson:"_id"`
+	}{}
+
+	for cursor.Next(ctx) {
+		err = cursor.Decode(&document)
+		if err == mongo.ErrNoDocuments {
+			continue
+		}
+
+		if err != nil {
+			_ = cursor.Close(ctx)
+			return errors.Wrap(err, "decoding GridFS metadata")
+		}
+
+		if b.opts.DryRun {
+			continue
+		}
+
+		if err = grid.Delete(document.ID); err != nil {
+			return errors.Wrap(err, "deleting GridFS file")
+		}
+	}
+	if err = cursor.Err(); err != nil {
+		return errors.Wrap(err, "iterating GridFS metadata")
+	}
+	if err = cursor.Close(ctx); err != nil {
+		return errors.Wrap(err, "closing cursor")
+	}
+
+	return nil
```

Comment:
It seems like the logic in `Remove` devolves into a call to `RemoveMany` with one key. Can we consolidate the two to reduce duplication?